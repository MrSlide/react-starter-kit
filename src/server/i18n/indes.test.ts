jest.mock('../constants/paths', function () {
  return {
    TRANSLATION_ASSETS_PATH: '/translations'
  }
})

jest.mock('../utils/fs', function () {
  return {
    find: function () {
      return [
        'en/test-a.json',
        'en/test-b.json',
        'en-gb/test-a.json',
        'pt-pt/test-a.json',
        'pt-pt/test-b.json'
      ]
    },
    readFile: async function (asset) {
      switch (asset) {
        case '/translations/en/test-a.json':
          return JSON.stringify({
            greeting: 'Hello, %{name}',
            test: 'This is a test'
          })
        case '/translations/en/test-b.json':
          return JSON.stringify({
            foo: 'Foo',
            bar: 'Bar'
          })
        case '/translations/en-gb/test-a.json':
          return JSON.stringify({
            greeting: 'Good day, %{name}'
          })
        case '/translations/pt-pt/test-a.json':
          return JSON.stringify({
            greeting: 'Olá, %{name}',
            test: 'Isto é um teste'
          })
        case '/translations/pt-pt/test-b.json':
          return JSON.stringify({
            foo: 'Foo',
            bar: 'Bar'
          })
      }
    }
  }
})

describe('Server side i18n', function () {
  const enabledLangs = ['en', 'en-GB']
  let i18n

  beforeEach(function () {
    process.env.CONFIG = JSON.stringify({
      localization: {
        enabledLangs,
        langMapping: {
          'en-US': 'en',
          'pt-BR': 'pt-PT'
        }
      }
    })

    i18n = require('.')
  })

  describe('init()', function () {
    test('loads the phrases', async function () {
      expect(i18n.getPhrases('en')).toBeUndefined()
      expect(i18n.getPhrases('en-GB')).toBeUndefined()

      await i18n.init()

      expect(i18n.getPhrases('en')).toMatchSnapshot()
      expect(i18n.getPhrases('en-GB')).toMatchSnapshot()
    })
  })

  describe('getAvailableLanguages()', function () {
    beforeEach(async function () {
      await i18n.init()
    })

    test('returns the loaded languages', function () {
      const output = i18n.getAvailableLanguages()

      expect(output).toEqual(enabledLangs)
    })
  })

  describe('getPhrases()', function () {
    beforeEach(async function () {
      await i18n.init()
    })

    test('returns the phrases for a language', function () {
      const output = i18n.getPhrases('en-GB')

      expect(output).toMatchSnapshot()
    })

    test('returns undefined if a language is not available', function () {
      const output = i18n.getPhrases('en-PT')

      expect(output).toBeUndefined()
    })
  })

  describe('getT()', function () {
    beforeEach(async function () {
      await i18n.init()
    })

    test('returns a t function for a specified language', function () {
      const t = i18n.getT('en-GB')

      expect(t('testA.greeting', { name: 'World' })).toBe('Good day, World')
    })

    test('returns the same t function for the same language', function () {
      const tA = i18n.getT('en-GB')
      const tB = i18n.getT('en-GB')

      expect(tA).toBe(tB)
    })

    test('throws an error if a specified language is not available', function () {
      expect(function () {
        i18n.getT('pt-PT')
      }).toThrowErrorMatchingSnapshot()
    })
  })
})
