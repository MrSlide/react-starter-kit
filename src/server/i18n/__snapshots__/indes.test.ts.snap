// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server side i18n getPhrases() returns the phrases for a language 1`] = `
Object {
  "testA": Object {
    "greeting": "Good day, %{name}",
    "test": "This is a test",
  },
  "testB": Object {
    "bar": "Bar",
    "foo": "Foo",
  },
}
`;

exports[`Server side i18n getT() throws an error if a specified language is not available 1`] = `"The language 'pt-PT' is not available."`;

exports[`Server side i18n init() loads the phrases 1`] = `
Object {
  "testA": Object {
    "greeting": "Hello, %{name}",
    "test": "This is a test",
  },
  "testB": Object {
    "bar": "Bar",
    "foo": "Foo",
  },
}
`;

exports[`Server side i18n init() loads the phrases 2`] = `
Object {
  "testA": Object {
    "greeting": "Good day, %{name}",
    "test": "This is a test",
  },
  "testB": Object {
    "bar": "Bar",
    "foo": "Foo",
  },
}
`;
